JSON logging

hcarty opened the pull request:
This adds JSON log output to the synchronous and Lwt log configurations, as well as an option to select between the two from the command line.  It also moves and renames things, so the interface breaks.

JSON output includes:
- Log message
- Log level
- Log source
- Tags

hcarty committed:
Bump ocamlformat version

hcarty committed:
Add JSON logging support

hcarty committed:
Add cmdliner terms for CLI selection of log format

pjhampton commented on src_lwt/ezlogs_cli_lwt.ml:
```diff
@@ -1,52 +1,107 @@
+let setup (format : format) style_renderer level =
+  match format with
+  | Line -> Line.setup style_renderer level
+  | Json -> Json.setup level
```
Line + Json 🤯 
That's awesome!

pjhampton commented on src/ezlogs_cli.ml:
```diff
@@ -1,22 +1,94 @@
+                (("level", `String (Logs.level_to_string (Some level)))
+                :: ("source", `String (Logs.Src.name src))
+                :: ("message", `String message)
+                :: tags_dict
```
Just to check, does this include a timestamp and other ECS base level fields?
https://www.elastic.co/guide/en/ecs/current/ecs-base.html

hcarty replied:
Very good call re: including ECS base level fields.
- `@timestamp`: **EDIT** After reading the ECS docs more properly, we should include this.  I'll make that change.
- `labels`: We can drop `tags` directly in this if we want, though that doesn't really seem like a perfect fit.  Maybe we ship a pre-defined [tag definition](https://erratique.ch/software/logs/doc/Logs/Tag/index.html#type-def) for labels?
- `message`: This is in
- `tags`: Unfortunate naming conflict with `Logs`'s concept of tags, but oh well.  Similar to labels, we could/should ship a `Logs` tag definition for this.

pjhampton commented on src/ezlogs_cli.ml:
```diff
@@ -1,22 +1,94 @@
+            let json : Json.t =
+              `Assoc
+                (("level", `String (Logs.level_to_string (Some level)))
+                :: ("source", `String (Logs.Src.name src))
```
I'm not sure `source` is the best as it conflicts with a filebeat metafield.

hcarty replied:
We can store this under ECS `labels` or `tags`?  I'm not sure which is correct.

stevewritescode replied:
`labels` would be better since we'd want this to be a key/value pair (e.g "source: foo"). `tags` is for a flat array of values. That said, we may want to use `service.name` instead of `source`

hcarty replied:
Changed to `log.logger` per ECS - thank you both!

pjhampton commented on src/ezlogs_cli.ml:
```diff
@@ -1,22 +1,94 @@
+              `Assoc
+                (("level", `String (Logs.level_to_string (Some level)))
+                :: ("source", `String (Logs.Src.name src))
+                :: ("message", `String message)
```
Please events :) https://www.elastic.co/guide/en/ecs/current/ecs-event.html
I would love to have an event / alert in the log like `BenignerFoundUniqueExample` so I could aggregate over them

pjhampton replied:
Maybe also log level in a separate field?

hcarty replied:
The log level is here, unless you meant something different? https://github.com/hcarty/ezlogs-cli/pull/1/files#diff-5ab8778f2f1fd8509c6ec40348d088cbR53

For events, I think we can capture them with [tags](https://erratique.ch/software/logs/doc/Logs/Tag/index.html).  It's worth having convenient support for that but I'm not sure if that should live in this library or somewhere else.  If it lives in this library then we should probably change the name since it's getting away from just CLI parsing + logging setup.

Expanding scope + renaming would not be a bad thing since we haven't made a proper opam release yet.  And we if we make it ECS-ready and APM-ready then that seems well worth expanding into.

hcarty replied:
Moved log level to `log.level` for more ECSing.  Thanks!

pjhampton commented:
This rocks 🤘

I'm still not 100% sure I understand this ECS business but others in team like it and going down this route. It would be awesome to enable it for EZlogs if required.

hcarty committed:
Make fields more ECS compliant

hcarty committed:
Add simple examples to README.md

This uses mdx for the README updates and refactors a bit of internal
code to make the README examples easier to write... which probably means
they're useful changes for users as well.

hcarty committed:
Merge remote-tracking branch 'origin/master' into json-logging

hcarty committed:
Move packaging info to dune-project

hcarty commented:
Tests are failing because of the timestamps in `README.md` 🤦   I'll think some more about how to avoid that.

https://github.com/hcarty/ezlogs-cli/pull/1/checks?check_run_id=714990596#step:9:26

hcarty committed:
Support passing in timestamp via epoch tag

hcarty commented:
Timestamps/epochs can now be passed in via tags - updated the README to use that for more README stability.

hcarty commented:
Windows CI failures are due to some bad interaction between mdx and Windows (and maybe dune?).

hcarty commented:
I'm going to merge this then work on a follow-up PR with ECS-ready tag definitions.

The Windows failures are due to mdx not handling Windows executables' `.exe` extensions properly.  I submitted a PR to hopefully fix that - https://github.com/realworldocaml/mdx/pull/263
